### ISSUE TITLE
abstract hashing algorithm to cluster initialization

### ISSUE THREAD

**Issue by duxing** [pfreixes/emcache#41]
as of version `0.3.0b0`, it seems like the hashing algorithm is not configurable ([reference](https://github.com/pfreixes/emcache/blob/675df6774b0e771d1a86c504da9f53cb08e15911/emcache/cluster.py))

I think it will help with adoption if hash algorithm can be configured. I'm writing a new application that need to read from an existing memcached cluster (written to by a component who's using CRC hash algorithm). Not being able to configure hash algorithm basically killed `emcache` for the application I'm working on, and this can easily become a blocker for other people working on existing applications too.

`emcache` is very well written with extreme performance and I think improvements in compatibility will significantly improve the adoption

**Comment by pfreixes** [comment]
Sorry for being late on this !!!

> I think it will help with adoption if hash algorithm can be configured. I'm writing a new application that need to read from an existing memcached cluster (written to by a component who's using CRC hash algorithm). Not being able to configure hash algorithm basically killed emcache for the application I'm working on, and this can easily become a blocker for other people working on existing applications too.

Does this mean that you would like to have a CRC eligible option for picking the right node? as it is being done here [1]

If so I'm open to adding other hashing algorithms for selecting the node distribution, but they would need to be provided by Emcache, so during the client creation you could choose between hashing algorithm and using by default the current `murmur3`

Is this what you are asking for?

[1] https://github.com/pfreixes/emcache/blob/675df6774b0e771d1a86c504da9f53cb08e15911/emcache/_cython/hashing/rendezvous.pyx#L58

**Comment by duxing** [comment]
thanks for getting back to me.

yes that's pretty much what I was looking for. I end up creating an array of clients (one mc host per client) and  implemented CRC hash used for node locating on my application level.

```
# ((crc32(<key>) >> 16) & 0x7fff) % <cluster_size>
# reference: https://github.com/couchbase/spymemcached/blob/master/src/main/java/net/spy/memcached/DefaultHashAlgorithm.java#L107
```

I think it'll be helpful for people to adopt this library for their existing projects. not a contract-breaking issue.

**Comment by pfreixes** [comment]
Looking on what you did by adding the CRC as an alternative hash function would not be enough. Emcache uses a Rendezvous hashing algorithm for distributing the keys around the configured Memcached servers, by adding the option of using a CRC hash method rather than using the murmur3 this will still be using the Rendezvous hashing algorithm, which won't guarantee having the keys in the nodes that you are expecting since you used the module function.

To fully supporting what you are asking for we should provide not only a way for configuring what hash algorithm is used under the hood but also supporting a way for choosing what routing algorithm would be used.

Im still thinking that is doable and could be useful for the users. Would be up to the user to define what distribution algorithm and hash function will be used behind the scenes, having for example the following combinations:

- Module + CRC
- Module + murmur3
- Rendezvous + CRC
- Rendezvous + murmur3 (would be the default one)

Then it would be up to the user for selecting what is best for the user, considering that:

- Rendezvous would perform worst when tens of nodes are used - all hashes for all nodes have to be calculated - but would reduce the total number of keys re sharded when a node is down or a new node is added into the pool. With Module strategy, all keys would be re sharded in case of having a node down or adding a new node, but performance would be better for picking up the right node.
- Murmur3 should provide you a better distribution key but with worse performance.

**Comment by pfreixes** [comment]
I would like to address this by providing an interface for overriding the default implementation, so during the factory of the `create_client` you would have the chance of providing an alternative implementation for picking one or another node.

But my plans for now are basically:

- Support SSL
- Support binary protocol
- Support SASL

But external contributions are more than wellcomed.

**Comment by duxing** [comment]
@pfreixes yeah that will be really nice to have. I would have picked the same approach if I were you.
amazing work btw. I just wish more ppl / projects can adopt this lib.

**Comment by ArtemIsmagilov** [comment]
https://en.wikipedia.org/wiki/Rendezvous_hashing
